go/pkg/libproxy/multiplexed.go

```diff
@@ -424,14 +410,22 @@ func (m *multiplexer) appendEvent(e *event) {
 	m.events = m.events.Next()
 }
 
-func (m *multiplexer) send(f *Frame) error {
+// send a frame (header) plus optional payload. If this call fails then the multiplexed connection will be desynchronised.
+func (m *multiplexer) send(f *Frame, payload []byte) error {
 	m.writeMutex.Lock()
 	defer m.writeMutex.Unlock()
+	m.appendEvent(&event{eventType: eventSend, frame: f})
+
 	if err := f.Write(m.connW); err != nil {
-		return err
+		return fmt.Errorf("writing frame %s: %w", f, err)
 	}
-	m.appendEvent(&event{eventType: eventSend, frame: f})
-	return m.connW.Flush()
+	if n, err := m.connW.Write(payload); err != nil || n != len(payload) {
```

Comment:
Shout it return n?

Reply:
Can you write nil?

Reply:
From friday's discussion: the [Write method docs](https://pkg.go.dev/io#Writer) has
> Write writes len(p) bytes from p to the underlying data stream.

So since `len(nil) == 0` it makes the call a no-op.

also
> Write must return a non-nil error if it returns n < len(p).

So it will return an error to the caller if `n` is too small.

Does this still sound reasonable?

Reply:
yes!